File: apps/roost-cli/src/state.ts
// `roost state` — generate STATE.md content. Called by Claude Code's
// Stop hook (.claude/settings.json) to auto-update STATE.md after
// each session. Writes to stdout; hook redirects to STATE.md.
// REWRITE.md R0.11.

import { spawn } from "bun";

// Resolve repo root from this file's location: apps/roost-cli/src/state.ts
// → ../../../ is the repo root regardless of where the CLI is invoked from.
const REPO_ROOT = new URL("../../../", import.meta.url).pathname;

async function out(cmd: string[]): Promise<string> {
  const p = spawn({ cmd, stdout: "pipe", stderr: "pipe", cwd: REPO_ROOT });
  const text = await new Response(p.stdout).text();
  await p.exited;
  return text.trimEnd();
}

export async function state(_args: string[]): Promise<void> {
  const branch = await out(["git", "branch", "--show-current"]);
  const commits = await out(["git", "log", "--oneline", "-5"]);
  const status = await out(["git", "status", "--short"]);
  const dateUtc = new Date().toISOString();

  // Truncate noisy status when long
  const statusBlock = status.split("\n").slice(0, 30).join("\n");

  const md = [
    "<!-- AUDIENCE: claude (auto-updated by Stop hook per R0.11) -->",
    "# STATE — Roost v2 rewrite snapshot",
    "",
    `updated=${dateUtc}`,
    `branch=${branch}`,
    "",
    "## last 5 commits",
    "```",
    commits,
    "```",
    "",
    "## git status",
    "```",
    statusBlock || "(clean)",
    "```",
    "",
    "## next action",
    "Past R4.5 cutover; `apps_legacy/` deleted (phase-24g). No queued spine-phase.",
    "Active work tracked per-commit (`phase-<slug>:`). See `git log --oneline` for current arc.",
    "Reopen REWRITE.md with a new R-anchor (R11+) before the next architectural shift.",
    "",
  ].join("\n");
  console.log(md);
}

File: crates/roost-cli/src/ops/state.rs
//! `roost state` — print the STATE.md snapshot an agent's Stop hook redirects
//! into the repository. Called by the crate's dispatcher. Reads git and
//! nothing else: every fact in the snapshot is a fact about the checkout, and a
//! command that also reported service health would have to keep two truths in
//! step.
//!
//! One thing genuinely changed from the TypeScript original, and it is worth
//! naming: it resolved the repository root from its own source path
//! (`new URL("../../../", import.meta.url)`), which a compiled binary does not
//! have. The root is an argument here, defaulting to the working directory.

use std::path::PathBuf;
use std::process::ExitCode;

use clap::Args;

use crate::command_error::CommandFailure;
use crate::utc_clock::format_utc_minute;
use crate::wall_clock;

/// How many `git status` lines the snapshot carries. Past this it is a wall of
/// unrelated filenames and the block stops being a summary.
pub const STATE_STATUS_LINES: usize = 30;

#[derive(Debug, Args)]
#[command(about = "Print a STATE.md snapshot of this checkout")]
pub struct StateArgs {
    #[arg(long, value_name = "DIR", default_value = ".")]
    pub repo: PathBuf,
}

pub fn run(args: &StateArgs) -> Result<ExitCode, CommandFailure> {
    let branch = git(&args.repo, &["branch", "--show-current"])?;
    let commits = git(&args.repo, &["log", "--oneline", "-5"])?;
    let status: String = git(&args.repo, &["status", "--short"])?
        .lines()
        .take(STATE_STATUS_LINES)
        .collect::<Vec<_>>()
        .join("\n");
    let status_block = if status.is_empty() {
        "(clean)"
    } else {
        &status
    };
    print!(
        "<!-- AUDIENCE: claude (auto-updated by Stop hook) -->\n\
         # STATE — Roost v3 snapshot\n\
         \n\
         updated={}\n\
         branch={branch}\n\
         \n\
         ## last 5 commits\n\
         ```\n\
         {commits}\n\
         ```\n\
         \n\
         ## git status\n\
         ```\n\
         {status_block}\n\
         ```\n\
         \n\
         ## next action\n\
         Active work is tracked per-commit (`<area>: <scope>`). See `git log --oneline` for\n\
         the current arc, and record the next architectural shift in ARCHITECTURE.md\n\
         before starting it.\n",
        format_utc_minute(wall_clock::now_ms())
    );
    Ok(ExitCode::SUCCESS)
}

fn git(repo: &std::path::Path, args: &[&str]) -> Result<String, CommandFailure> {
    let output = std::process::Command::new("git")
        .current_dir(repo)
        .args(args)
        .output()
        .map_err(|error| CommandFailure::generic(format!("git {}: {error}", args.join(" "))))?;
    if !output.status.success() {
        return Err(CommandFailure::generic(format!(
            "git {} failed in {}",
            args.join(" "),
            repo.display()
        )));
    }
    Ok(String::from_utf8_lossy(&output.stdout)
        .trim_end()
        .to_string())
}
